File: Test_Ball_Detection/DetectBall.py
import sys
import time
import cv2
import numpy as np
#import Image
import random
import math
import os
import warnings
warnings.filterwarnings("ignore")

def detect_ball_image():
    N = 0
    L = ["ball_"+str(i) for i in range(1,9)]
    ecc_l = []
    is_it = []
    image_name = []
    radius_l = []
    for x in L :
        x  = x + '.png'
        image_name+=[x]

        # read image from file
        cvImg = cv2.imread(x)
        cvImg_blur = cv2.GaussianBlur(cvImg, (11,11),0)
        # Get the HSV image
        cvImg_hsv = cv2.cvtColor(cvImg_blur, cv2.COLOR_BGR2HSV)
        # get dimensions
        imageHeight, imageWidth, imageChannels = cvImg.shape
    

        # Find the ornage ball
        lower_orange = (0,80,160)
        upper_orange = (13,255,255)
    
        mask = cv2.inRange(cvImg_hsv, lower_orange, upper_orange)
        mask = cv2.erode(mask, None, iterations=2)
        mask = cv2.dilate(mask, None, iterations=2)
    
    
        # Find the contours of the orange ball
        ret, thresh = cv2.threshold(mask, 127,255,0)
        im2, contours, hierarchy = cv2.findContours(thresh, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
        M = cv2.moments(thresh)
        if M["m00"] == 0:
            cX, cY = 0, 0
        else : 
            # calculate x,y coordinate of center
            cX = int(M["m10"] / M["m00"])
            cY = int(M["m01"] / M["m00"])
        center = (cX, cY)
        print(center)
        length = len(contours)
        d = 100
        radius = 100
        if length > 0 : 
            cont = np.vstack(contours[i] for i in range(length))
      
            cnt_len = cv2.arcLength(cont, True)
            cont = cv2.approxPolyDP(cont, .01*cnt_len, True)
            hull = cv2.convexHull(cont)
      
                  
            uni_hull = []
            uni_hull.append(hull) # <- array as first element of list
            #cv2.drawContours(cvImg,contours,-1,(0, 255, 0),2)
            # calculate points for each contour
            if len(hull) > 4 :
                ellipse = cv2.fitEllipse(hull)
                
                cv2.ellipse(cvImg, ellipse, (0,0,255), thickness = 2)

                radius = 0
                if (len(contours) > 0) :
                    x, y, w, h = cv2.boundingRect(contours[0])                    
                    ellipse = cv2.fitEllipse(contours[0])
                    cv2.rectangle(cvImg,(x,y),(x+w,y+h),(0,255,0),3)
                    area = cv2.contourArea(contours[0])
                    radius = math.sqrt(area* 2.54/300)/(math.pi) #cm
                    d = abs(h - w)
                    print("ecc" , d)
                    if d < 4:
                        print("It is a ball!")
                        is_it+=["It is a ball!"]
                    else:
                        print("It is not a ball!")
                        is_it+=["It is not a ball!"]

                print("radius", radius)
                # create a window
        
                cv2.namedWindow("Nao Image")
                # show the image
                cv2.imshow("Nao Image",cvImg)
        
                # wait until a key is pressed
                cv2.waitKey(0)
                cv2.destroyWindow("Nao Image")
                N+=1
            else:
                print("No contour found")
                is_it+=["No contour found"]

        else:
            print("No contour found")
            is_it+=["No contour found"]
        ecc_l+= [d]
        radius_l+=[radius]
    print(N)
    for i in range(len(is_it)):
        print(image_name[i])
        print(ecc_l[i])
        print(is_it[i])
        print(radius_l[i])
    cv2.destroyAllWindows()


detect_ball_image()
#print(detect_ball(cv2.imread("ball_3.png")))

File: DetectBall.cpp
// Prevent Visual Studio Intellisense from defining _WIN32 and _MSC_VER when we use 
// Visual Studio to edit Linux or Borland C++ code.
#ifdef __linux__
#	undef _WIN32
#endif // __linux__
#if defined(__GNUC__) || defined(__BORLANDC__)
#	undef _MSC_VER
#endif // defined(__GNUC__) || defined(__BORLANDC__)

#include "DetectBall.h"


/*
    File created for the SAILBOAT Project in Guerledan
    Used for defining the function for navigating with the Sailboat
*/

void detect_ball(IplImage* image, IplImage* overlayimage, bool ballFound){
    cv::Mat initImage = cv::cvarrToMat(image, true);
    cv::Mat gaussianImage = cv::cvarrToMat(image, true);
    cv::Mat hSVImage = cv::cvarrToMat(image, true);
    cv::Mat mask, thresh, hull;
    ovlImage = cv::cvarrToMat(overlayimage, false);
    cv::Mat element = cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size( 3, 3 ));
    cv::Moments oMoments;
    vector <vector <Point> > contours;
    vector <Point> cont;
    cv::Mat cnt;
    double cnt_len, radius, aArea, dM01, dM10, dM00;
    cv::RotatedRect eEllipse;
    cv::Rect rRect;
    int hull_length;

    ballFound = false;
    // Apply Gaussian Blur
    cv::GaussianBlur( initImage, gaussianImage, cv::Size(11, 11), 0, 0 );
    
    // Get the HSV Images
    cv::cvtColor(gaussianImage, hSVImage, CV_BGR2HSV);
    
    // Create Mask
    cv::inRange(hSVImage, Scalar(LOWER_H, LOWER_S, LOWER_V), Scalar(UPPER_H, UPPER_S, UPPER_V), mask);
    /// Apply Mask
    cv::dilate(mask, mask, element, Point(-1,-1), 2);
    cv::erode(mask, mask, element, Point(-1,-1),  2);

    // Find contours of ball
    cv::threshold(mask, thresh, 127, 255, 0);
    cv::findContours(thresh, contours, RETR_TREE, CHAIN_APPROX_SIMPLE);

    // Compute mandatory moments
    oMoments = cv::moments(thresh);
    dM01 = oMoments.m01;
    dM10 = oMoments.m10;
    dM00 = oMoments.m00;

    // Compute x,y coordinate of center
    ball_x = (dM00 == 0) ?  0 : (int) (dM10/dM00);
    ball_y = (dM00 == 0) ?  0 : (int) (dM01/dM00);
    
    // 0.0002645833 for converting pixels to meters
    ball_distance = (double) sqrt(pow(ball_x*0.0002645833, 2) + pow(ball_y*0.0002645833, 2));
    // Some contours are found, let's process them
    if((int) contours.size() != 0){
        // Stack the contours
        for(int i=0; i < contours.size(); i++){
            cont.insert(cont.end(), contours[i].begin(), contours[i].end());
        }
        
        // Approximate them as a ellipse-like shape. 
        cnt_len = cv::arcLength(cont, true);
        cv::approxPolyDP((cv::Mat)cont, cnt, 0.01*cnt_len, true);
        
        cv::convexHull(cnt, hull);

        hull_length =  (int) hull.rows * hull.cols;
        // If we have enough points in the shape, process it 
        if(hull_length > 5){
            // Shape it as an ellipse
	        fflush(stdout);
            eEllipse = cv::fitEllipse(hull);
            fflush(stdout);

            // Draw the ellipse
            cv::ellipse(ovlImage, eEllipse, Scalar(ELLIPSE_COLOR_R, ELLIPSE_COLOR_G, ELLIPSE_COLOR_B));
            ball_radius = 0;

            // Draw the best bouding rectangle
            rRect = cv::boundingRect(hull);
            cv::rectangle(ovlImage, rRect, Scalar(RECTANGLE_COLOR_R, RECTANGLE_COLOR_G, RECTANGLE_COLOR_B));
            aArea = cv::contourArea(hull);

            // Find the radius of the shape (we assume it is almost a circle, though it is not always true)
            ball_radius = sqrt((float) (aArea * pow(0.0002645833,2) /M_PI));

            // Assert the object is a ball. Tried with the ellipse eccentricity but it gave me absurd result
            // So I made an approximation with the properties of the bouding rectangle. (Almost work)
            ballFound = (abs((long) (rRect.height - rRect.width)) < 40) ? true : false;

        }
    }
}